overlord/configstate/config/transaction_test.go

```diff
@@ -604,3 +606,274 @@ func (s *transactionSuite) TestPristineGet(c *C) {
+		n++
+
+		s := fmt.Sprintf("%s=virtual-value", key)
+		return s, nil
+	})
+	c.Assert(err, IsNil)
+
+	tr := config.NewTransaction(s.state)
+
+	var res map[string]interface{}
+	// the root doc
+	err = tr.Get("some-snap", "", &res)
+	c.Assert(err, IsNil)
+	c.Check(res, DeepEquals, map[string]interface{}{
+		"some-key":  "some-value",
+		"other-key": "value",
+		"key": map[string]interface{}{
+			"virtual": "key.virtual=virtual-value",
+		},
+	})
+}
+
+func (s *transactionSuite) TestVirtualGetSubtreeMerged(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+	s.state.Set("config", map[string]map[string]interface{}{
+		"some-snap": {
+			"other-key": "other-value",
+			"real-and-virtual": map[string]interface{}{
+				"real": "real-value",
+			},
+		},
+	})
+
+	n := 0
+	err := config.RegisterVirtualConfig("some-snap", "real-and-virtual.virtual", func(key string) (interface{}, error) {
+		n++
+
+		s := fmt.Sprintf("%s=virtual-value", key)
+		return s, nil
+	})
+	c.Assert(err, IsNil)
+
+	tr := config.NewTransaction(s.state)
+
+	var res string
+	// non-virtual keys work fine
+	err = tr.Get("some-snap", "other-key", &res)
+	c.Assert(err, IsNil)
+	c.Check(res, Equals, "other-value")
+	// no virtual helper was called because the requested key was not
+	// part of the virtual configuration
+	c.Check(n, Equals, 0)
+
+	var res2 map[string]interface{}
+	err = tr.Get("some-snap", "real-and-virtual", &res2)
+	c.Assert(err, IsNil)
+	c.Check(res2, HasLen, 2)
+	// real
+	c.Check(res2["real"], Equals, "real-value")
+	// and virtual values are combined
+	c.Check(res2["virtual"], Equals, "real-and-virtual.virtual=virtual-value")
+	// the virtual config function was called
+	c.Check(n, Equals, 1)
+}
+
+func (s *transactionSuite) TestVirtualCommitValuesNotStored(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	err := config.RegisterVirtualConfig("some-snap", "key.nested.virtual", func(key string) (interface{}, error) {
+		c.Errorf("virtual func should not get called in this test")
+		return nil, nil
+	})
+	c.Assert(err, IsNil)
+	err = config.RegisterVirtualConfig("some-snap", "simple-virtual", func(key string) (interface{}, error) {
+		c.Errorf("virtual func should not get called in this test")
+		return nil, nil
+	})
+	c.Assert(err, IsNil)
+
+	tr := config.NewTransaction(s.state)
+
+	// set some values
+	c.Check(tr.Set("other-snap", "key", "value"), IsNil)
+	c.Check(tr.Set("some-snap", "key.nested.virtual.sub", "won't-get-set"), IsNil)
+	c.Check(tr.Set("some-snap", "key.nested.virtual.sub2.sub2sub", "also-won't-get-set"), IsNil)
+	c.Check(tr.Set("some-snap", "key.not-virtual", "value"), IsNil)
+	c.Check(tr.Set("some-snap", "key.nested.not-virtual", "value"), IsNil)
+	// top level virtual with no nesting
+	c.Check(tr.Set("some-snap", "simple-virtual", "will-not-get-set"), IsNil)
+	tr.Commit()
+
+	// and check what got stored in the state
+	var config map[string]map[string]interface{}
+	s.state.Get("config", &config)
```

Comment:
this test is not testing what it should now, it should check for the len of config["some-snap"] itself which should be 1 but is 2 afaict, the clear config logic needs changes to deal with this case as I suspected

Reply:
Uh, nice catch. Thanks, I updated the tests and the code.